### PR TITLE
Complete queue push and action class and command line.

### PR DESCRIPTION
**Pull request by benbowler**
Services the following Trello cards:

https://trello.com/c/k5iqpGkO/32-1-adding-single-posts-to-queue-1
https://trello.com/c/9E4nSVpV/31-8-choose-library-autoload-working-in-code-and-connecting-to-queue-to-handle-queue-illuminate-laravel-5
https://trello.com/c/qzOKRiPv/22-3-make-the-import-functions-available-via-cli
https://trello.com/c/SSZ05ESN/33-5-adding-multiple-posts-to-queue
https://trello.com/c/VT5nbif4/21-5-update-catfish-importer-interface-to-make-use-of-the-new-queue-functions-5
https://trello.com/c/RMH5WlJc/35-5-command-line-agent-that-process-items-from-the-queue-and-logs-success-or-error-13
https://trello.com/c/3rqNCcGx/34-1-flush-the-queue

In future I'll split pull requests.

### CODE REVIEW

File: app/Services/Worker.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  //         $this->handleJobException($connection, $job, $delay, $e);
+  //     } catch (Throwable $e) {
+  //         $this->handleJobException($connection, $job, $delay, $e);
+  //     }
+  // }
+
+  /**
+   * Purge queue function for tests primarily
+   */
+  public function purge($connection, $queue)
```
Comment (ecoad) [inline app/Services/Worker.php]: Just pop the opening brace on this line

File: composer.json [hunk 2]
```diff
@@ -12,7 +12,11 @@
     "maknz/slack": "^1.7",
     "sabre/xml": "^1.4",
     "jarednova/mesh": "0.4",
-    "shortlist-digital/agreable-base-theme": "~2.1"
+    "shortlist-digital/agreable-base-theme": "~2.1",
+    "illuminate/queue": "5.2.*",
+    "illuminate/encryption": "5.2.*",
+    "aws/aws-sdk-php": "dev-master",
```
Comment (ecoad) [inline composer.json]: Can you pick a tag from here? https://packagist.org/packages/aws/aws-sdk-php how about `3.23.*`?

File: composer.json [hunk 3]
```diff
@@ -12,7 +12,11 @@
...
+    "wp-cli/wp-cli": "dev-master"
```
Comment (ecoad) [inline composer.json]: https://packagist.org/packages/wp-cli/wp-cli `1.1.*`?

File: composer.json (hunk 2 above)
Comment (ecoad) [inline composer.json]: (We tend to dislike `dev-master` as its always bloody updating, and doesn't lock down breaking changes)

File: tests/features/bootstrap/SyncContext.php [hunk 4]
```diff
@@ -10,51 +18,91 @@
...
+  // Protected variables for the tests to use
+  protected static $queueID;
   protected static $categorySitemap;
   protected static $totalStatus;
 
   /**
-   * @Given /^I sync (\d+) most recent posts from the category sitemap "([^"]*)"$/
+   * @Given /^I purge the queue$/
+   */
+  public function iPurgeTheQueue()
```
Comment (ecoad) [inline tests/features/bootstrap/SyncContext.php]: Pop the opening curly brace on this line, ta. Check the other functions below too.

File: tests/features/bootstrap/SyncContext.php [hunk 5]
```diff
@@ -72,5 +120,19 @@ public function iShouldSeeACoupleOfImportedOutOfAtLeast($arg) {
     Assert::assertGreaterThan(10000, self::$totalStatus->total);
   }
 
+  /**
+   * @Given /^I retrive the "([^"]*)" category import status$/
```
Comment (ecoad) [inline tests/features/bootstrap/SyncContext.php]: Typo on `retrive`

File: app/Services/Sync.php [hunk 6]
```diff
@@ -3,39 +3,210 @@
...
+        // delay
+        // time before retries
+        // max number of tries
+
+        $worker->pop('default', getenv('AWS_SQS_QUEUE'), 0, 3, 0);
+      } catch (Exception $e) {
+        if($cli) {
+          WP_CLI::error('Error processing next queue item.');
+        }
+        throw new Exception("Error processing next queue item.", 1); // $e
```
Comment (benbowler) [inline app/Services/Sync.php]: ✔

File: app/Services/Sync.php [hunk 7]
```diff
@@ -3,39 +3,210 @@
...
+    // Connect to the AWS SQS Queue
+    $worker = new Worker($queue->getQueueManager());
+
+    // Pass cli status to worker class
+    if($cli) {
+      $worker->cli = true;
+    }
+
+    try {
+      // Parameters:
```
Comment (benbowler) [inline app/Services/Sync.php]: Thanks.

File: app/Services/Sync.php [hunk 8]
```diff
@@ -3,39 +3,210 @@
...
+   * Queue Single URL
+   */
+  public static function queueUrl($url, $onExistAction = 'update') {
+   // Push item into Queue
+   if(is_string(Queue::push('importUrl', array('url' => $url, 'onExistAction' => $onExistAction)))) {
+     // Return post url
+     return $url;
+   }
+   // If it failed return false
+   return false;
```
Comment (benbowler) [inline app/Services/Sync.php]: Added some clear error messages, show the error message in the errors and added a couple more try catch's to boot!

File: app/Services/Queue.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+   self::getContainer()->bind('request', function() {
+   	return new Illuminate\Http\Request();
+   });
+
+   // Connect with credentials from .env
+   self::addConnection([
+       'driver' => 'sqs',
+       'key'    => getenv('AWS_KEY'),
+       'secret' => getenv('AWS_SECRET'),
+       'queue'  => getenv('AWS_SQS_QUEUE'),
```
Comment (benbowler) [inline app/Services/Queue.php]: Updated to:
```
ILLUMINATE_ENCRYPTOR_KEY=

AWS_KEY=
AWS_SECRET=
AWS_SQS_CATFISH_IMPORTER_REGION=
AWS_SQS_CATFISH_IMPORTER_QUEUE=
```

File: app/Services/Queue.php [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+   * Setup queue connection variables on class instantiation
+   */
+  public function __construct() {
+
+   parent::__construct();
+
+   // Connect to the AWS SQS Queue
+   // Bind required subclasses
+   self::getContainer()->bind('encrypter', function() {
+   	return new \Illuminate\Encryption\Encrypter(getenv('APP_KEY'), 'AES-256-CBC');
```
Comment (benbowler) [inline app/Services/Queue.php]: Changed to ILLUMINATE_ENCRYPTOR_KEY and added exception to warn.

File: README.md [hunk 11]
```diff
@@ -1,4 +1,55 @@
...
+
+```
+wp catfish listen --debug
+```
+### Purge command
+
+The purge command **deletes all queue items**. Used to give tests a clean environment to work with.
+
+```
+wp catfish listen --debug
```
Comment (benbowler) [inline README.md]: Added back in the option to clear the queue as per catchup.

File: README.md [hunk 12]
```diff
@@ -1,4 +1,55 @@
...
+wp catfish queue all --debug
+wp catfish queue http://www.shortlist.com/sitemap/entertainment/48-hours-to.xml --debug
+```
+
+### Work command
+
+The work command actions one single item in the queue.
+
+```
+wp catfish listen --debug
```
Comment (benbowler) [inline README.md]: Fixed and added details on setting up the environment variables.

File: README.md [hunk 13]
```diff
@@ -1,4 +1,55 @@
 Agreable Catfish Importer Plugin
 ===============
 
-TBC
+For importing Catfish content in to Croissant
```
Comment (benbowler) [inline README.md]: Any time.

**Pull request merged**